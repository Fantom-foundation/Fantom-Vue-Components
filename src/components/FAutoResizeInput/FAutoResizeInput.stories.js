import FAutoResizeInput from './FAutoResizeInput.vue';
import FInput from '../FInput/FInput.vue';

export default {
    title: 'FAutoResizeInput',
    component: FAutoResizeInput,
};

export const Default = () => ({
    components: { FAutoResizeInput, FInput },
    template: `
        <div>
            <f-auto-resize-input><input type="text" aria-label="auto resizable input"></f-auto-resize-input> <br /><br />
            <f-auto-resize-input><f-input label="FInput" /></f-auto-resize-input>
        </div>
    `,
});

export const MinWidth = () => ({
    components: { FAutoResizeInput, FInput },
    template: `
        <div>
            <f-auto-resize-input min-width="48px"><input type="text" aria-label="auto resizable input"></f-auto-resize-input> <br /><br />
            <f-auto-resize-input min-width="48px"><f-input label="FInput" /></f-auto-resize-input>
        </div>
    `,
});

export const Value = () => ({
    components: { FAutoResizeInput, FInput },
    template: `
        <div>
            <f-auto-resize-input><input type="text" value="Lorem ipsum" aria-label="auto resizable input"></f-auto-resize-input> <br /><br />
            <f-auto-resize-input><f-input label="FInput" value="Lorem ipsum" /></f-auto-resize-input>
        </div>
    `,
});

export const ExtraWidth = () => ({
    components: { FAutoResizeInput, FInput },
    template: `
        <div>
            <f-auto-resize-input :extra-width="21"><input type="number" value="0" aria-label="auto resizable input"></f-auto-resize-input> <br /><br />
        </div>
    `,
});

export const Placeholder = () => ({
    components: { FAutoResizeInput, FInput },
    template: `
        <div>
            <f-auto-resize-input><input type="text" placeholder="Placeholder"></f-auto-resize-input> <br /><br />
            <f-auto-resize-input><f-input label="FInput" placeholder="Placeholder" /></f-auto-resize-input>
        </div>
    `,
});
