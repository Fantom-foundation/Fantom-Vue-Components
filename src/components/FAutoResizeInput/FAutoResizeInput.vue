<template>
    <span class="fautoresizeinput">
        <slot></slot>
        <span ref="twin" class="fautoresizeinput_hiddentwin" aria-hidden="true"></span>
    </span>
</template>

<script>
import { clearElement } from '@/utils/dom2.js';

export default {
    name: 'FAutoResizeInput',

    props: {
        /** Minimal input width */
        minWidth: {
            type: String,
            default: '',
        },
        /** Add this number of pixels to the resulting width */
        extraWidth: {
            type: Number,
            default: 5,
        },
    },

    created() {
        /** @type {HTMLInputElement} */
        this._eInput = null;
    },

    mounted() {
        this.init();
    },

    beforeDestroy() {
        this.freeInput();
    },

    methods: {
        init() {
            this.setInput();

            const eInput = this._eInput;

            if (eInput) {
                this.update();

                if (eInput.placeholder) {
                    this.resizeInput(eInput.placeholder);
                }
            }
        },

        update() {
            this.setTwinStyle();
            this.resizeInput();
        },

        /**
         * @param {string} text
         */
        resizeInput(text = '') {
            const eInput = this._eInput;
            const eTwin = this.$refs.twin;

            if (eInput) {
                clearElement(eTwin);
                eTwin.appendChild(document.createTextNode(text || eInput.value));
                this._eInput.style.width = eTwin.offsetWidth + this.extraWidth + 'px';
            }
        },

        setTwinStyle() {
            const eTwin = this.$refs.twin;
            const eInputStyle = window.getComputedStyle(this._eInput);
            const props = [
                'paddingLeft',
                'paddingRight',
                'fontFamily',
                'fontSize',
                'fontStyle',
                'fontWeight',
                'letterSpacing',
                'textTransform',
            ];

            props.forEach(prop => (eTwin.style[prop] = eInputStyle[prop]));
        },

        setInput() {
            this.freeInput();

            const eInput = this.$el.querySelector('input');

            if (eInput) {
                eInput.addEventListener('input', this.onInput);
                eInput.addEventListener('change', this.onInput);

                eInput.style.boxSizing = 'border-box';
                if (this.minWidth) {
                    eInput.style.minWidth = this.minWidth;
                }

                this._eInput = eInput;
            }
        },

        freeInput() {
            if (this._eInput) {
                this._eInput.removeEventListener('input', this.onInput);
                this._eInput.removeEventListener('change', this.onInput);
                this._eInput = null;
            }
        },

        onInput() {
            this.resizeInput();
        },
    },
};
</script>

<style lang="scss">
@use 'style';
</style>
