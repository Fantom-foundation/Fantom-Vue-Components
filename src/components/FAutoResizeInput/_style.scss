.fautoresizeinput {
    display: inline-block;
    //max-width: fit-content;

    /*input {
        width: 100%;
    }*/

    &_hiddentwin {
        position: fixed;
        top: -1000px;
        left: -1000px;
        //visibility: hidden;
        white-space: pre;
        box-sizing: border-box;
        //overflow: scroll;
    }
}
