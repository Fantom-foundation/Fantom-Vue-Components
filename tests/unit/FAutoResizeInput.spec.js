import { mount } from '@vue/test-utils';
import FAutoResizeInput from '@/components/FAutoResizeInput/FAutoResizeInput.vue';

let wrapper = null;

function destroyWrapper(wrapper) {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}

function createWrapper({ propsData = {}, slots = { default: '<input type="text" />' } } = {}) {
    return mount(FAutoResizeInput, {
        propsData,
        slots,
        // attachTo: document.body,
    });
}

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('FAutoResizeInput', () => {
    it('should render correctly', () => {
        wrapper = createWrapper({ slots: {} });

        expect(wrapper.element).toMatchSnapshot();
    });

    // offsetWidth doesn't work properly in this test enviroment :(
    /* it('should resize input element properly while typing', async () => {
        wrapper = createWrapper({ slots: { default: '<input type="text" />' } });

        const eInput = wrapper.find('input');

        await eInput.setValue('eee');
    });*/

    it('should set minimal width of the input element if `minWidth` prop is specified', () => {
        wrapper = createWrapper({ propsData: { minWidth: '20px' } });

        const eInput = wrapper.find('input');

        expect(eInput.element.style.minWidth).toBe('20px');
    });

    it('should add an additional width to the resulting width if the number of pixels `extraWidth` is specified', () => {
        wrapper = createWrapper({ propsData: { extraWidth: 10 } });

        const eInput = wrapper.find('input');

        expect(eInput.element.style.width).toBe('10px');
    });
});
