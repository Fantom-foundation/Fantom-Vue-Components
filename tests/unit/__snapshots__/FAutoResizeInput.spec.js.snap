// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAutoResizeInput should render correctly 1`] = `
<span
  class="fautoresizeinput"
>
   
  <span
    aria-hidden="true"
    class="fautoresizeinput_hiddentwin"
  />
</span>
`;
